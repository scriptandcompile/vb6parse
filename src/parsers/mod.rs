//! This module contains the file parsers for the Visual Basic 6 language.
//!
//! The major components of this module are:
//!
//! `VB6Stream` - The main parsing structure for VB6 files. It holds the
//! bytes for a file and provides basic methoda for iterating over the bytes as
//! well as tracking the line, column, and character position of the current byte.
//!
//! `VB6Project` - a structure for holding the parsed information from a VB6
//! project file.
//!
//! `VB6ClassFile` - a structure for holding the parsed information from a VB6
//! class file.
//!
//! `VB6FormFile` - a structure for holding the parsed information from a VB6
//! form file.
//!
//! `VB6ModuleFile` - a structure for holding the parsed information from a VB6
//! module file.
//!

mod header;
mod vb6stream;

pub mod class;
pub mod compilesettings;
pub mod form;
pub mod module;
pub mod objectreference;
pub mod project;
pub mod properties;
pub mod vb6;

pub use class::{
    FileUsage, MtsStatus, Persistance, VB6ClassFile, VB6ClassHeader, VB6ClassProperties,
    VB6ClassVersion,
};

pub use compilesettings::{
    BoundsCheck, CompilationType, FloatingPointErrorCheck, OverflowCheck, PentiumFDivBugCheck,
    UnroundedFloatingPoint,
};
pub use form::{resource_file_resolver, VB6FormFile};
pub use header::{VB6FileAttributes, VB6FileFormatVersion};
pub use module::VB6ModuleFile;

pub use properties::Properties;

pub use objectreference::VB6ObjectReference;

pub use project::{
    CompileTargetType, VB6Project, VB6ProjectClass, VB6ProjectModule, VB6ProjectReference,
};

pub use vb6::{is_english_code, vb6_parse};

pub use vb6stream::VB6Stream;
