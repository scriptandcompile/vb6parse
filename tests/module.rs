use vb6parse::parsers::VB6ModuleFile;

#[test]
fn artificial_life_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Artificial-life/Declarations.bas");

    let module_file = match VB6ModuleFile::parse("Declarations.bas".to_owned(), module_file_bytes) {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Declarations.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn game_physics_basic_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Game-physics-basic/Physics_Logic.bas");

    let module_file = match VB6ModuleFile::parse("Physics_Logic.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Physics_Logic.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn histograms_advanced_module_load() {
    let module_file_bytes =
        include_bytes!("./data/vb6-code/Histograms-advanced/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn histograms_basic_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Histograms-basic/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn levels_effect_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Levels-effect/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn map_editor_2d_module_load() {
    let module1_file_bytes = include_bytes!("./data/vb6-code/Map-editor-2D/Subs.bas");

    let module1_file = match VB6ModuleFile::parse("Subs.bas".to_owned(), module1_file_bytes) {
        Ok(module1_file) => module1_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Subs.bas' module file");
        }
    };

    let module2_file_bytes = include_bytes!("./data/vb6-code/Map-editor-2D/Declarations.bas");

    let module2_file = match VB6ModuleFile::parse("Declarations.bas".to_owned(), module2_file_bytes)
    {
        Ok(module2_file) => module2_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Declarations.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module1_file);
    insta::assert_yaml_snapshot!(module2_file);
}
