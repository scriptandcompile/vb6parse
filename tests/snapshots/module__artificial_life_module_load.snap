---
source: tests/module.rs
expression: module_file
---
name:
  - 34
  - 68
  - 101
  - 99
  - 108
  - 97
  - 114
  - 97
  - 116
  - 105
  - 111
  - 110
  - 115
  - 34
  - 13
  - 10
tokens:
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 65
      - 80
      - 73
      - 32
      - 99
      - 97
      - 108
      - 108
      - 115
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 83
      - 116
      - 97
      - 110
      - 100
      - 97
      - 114
      - 100
      - 32
      - 105
      - 109
      - 97
      - 103
      - 101
      - 32
      - 98
      - 108
      - 105
      - 116
      - 116
      - 105
      - 110
      - 103
      - 59
      - 32
      - 102
      - 97
      - 115
      - 116
      - 101
      - 114
      - 32
      - 116
      - 104
      - 97
      - 110
      - 32
      - 86
      - 66
      - 39
      - 115
      - 32
      - 105
      - 110
      - 116
      - 101
      - 114
      - 110
      - 97
      - 108
      - 32
      - 80
      - 97
      - 105
      - 110
      - 116
      - 80
      - 105
      - 99
      - 116
      - 117
      - 114
      - 101
      - 32
      - 109
      - 101
      - 116
      - 104
      - 111
      - 100
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - DeclareKeyword:
      - 68
      - 101
      - 99
      - 108
      - 97
      - 114
      - 101
  - Whitespace:
      - 32
  - FunctionKeyword:
      - 70
      - 117
      - 110
      - 99
      - 116
      - 105
      - 111
      - 110
  - Whitespace:
      - 32
  - Identifier:
      - 66
      - 105
      - 116
      - 66
      - 108
      - 116
  - Whitespace:
      - 32
  - LibKeyword:
      - 76
      - 105
      - 98
  - Whitespace:
      - 32
  - StringLiteral:
      - 34
      - 103
      - 100
      - 105
      - 51
      - 50
      - 34
  - Whitespace:
      - 32
  - LeftParanthesis:
      - 40
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 68
      - 115
      - 116
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 110
      - 87
      - 105
      - 100
      - 116
      - 104
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 110
      - 72
      - 101
      - 105
      - 103
      - 104
      - 116
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 83
      - 114
      - 99
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
      - 83
      - 114
      - 99
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 121
      - 83
      - 114
      - 99
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 100
      - 119
      - 82
      - 111
      - 112
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 83
      - 116
      - 97
      - 110
      - 100
      - 97
      - 114
      - 100
      - 32
      - 112
      - 105
      - 120
      - 101
      - 108
      - 32
      - 115
      - 101
      - 116
      - 116
      - 105
      - 110
      - 103
      - 59
      - 32
      - 109
      - 117
      - 99
      - 104
      - 32
      - 102
      - 97
      - 115
      - 116
      - 101
      - 114
      - 32
      - 116
      - 104
      - 97
      - 110
      - 32
      - 86
      - 66
      - 39
      - 115
      - 32
      - 105
      - 110
      - 116
      - 101
      - 114
      - 110
      - 97
      - 108
      - 32
      - 80
      - 83
      - 101
      - 116
      - 32
      - 97
      - 110
      - 100
      - 32
      - 80
      - 111
      - 105
      - 110
      - 116
      - 32
      - 109
      - 101
      - 116
      - 104
      - 111
      - 100
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - DeclareKeyword:
      - 68
      - 101
      - 99
      - 108
      - 97
      - 114
      - 101
  - Whitespace:
      - 32
  - FunctionKeyword:
      - 70
      - 117
      - 110
      - 99
      - 116
      - 105
      - 111
      - 110
  - Whitespace:
      - 32
  - Identifier:
      - 71
      - 101
      - 116
      - 80
      - 105
      - 120
      - 101
      - 108
  - Whitespace:
      - 32
  - LibKeyword:
      - 76
      - 105
      - 98
  - Whitespace:
      - 32
  - StringLiteral:
      - 34
      - 103
      - 100
      - 105
      - 51
      - 50
      - 34
  - Whitespace:
      - 32
  - LeftParanthesis:
      - 40
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - DeclareKeyword:
      - 68
      - 101
      - 99
      - 108
      - 97
      - 114
      - 101
  - Whitespace:
      - 32
  - FunctionKeyword:
      - 70
      - 117
      - 110
      - 99
      - 116
      - 105
      - 111
      - 110
  - Whitespace:
      - 32
  - Identifier:
      - 83
      - 101
      - 116
      - 80
      - 105
      - 120
      - 101
      - 108
      - 86
  - Whitespace:
      - 32
  - LibKeyword:
      - 76
      - 105
      - 98
  - Whitespace:
      - 32
  - StringLiteral:
      - 34
      - 103
      - 100
      - 105
      - 51
      - 50
      - 34
  - Whitespace:
      - 32
  - LeftParanthesis:
      - 40
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 99
      - 114
      - 67
      - 111
      - 108
      - 111
      - 114
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - ByteKeyword:
      - 66
      - 121
      - 116
      - 101
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 83
      - 105
      - 122
      - 101
      - 32
      - 111
      - 102
      - 32
      - 116
      - 104
      - 101
      - 32
      - 109
      - 97
      - 112
      - 59
      - 32
      - 116
      - 104
      - 101
      - 115
      - 101
      - 32
      - 100
      - 105
      - 109
      - 101
      - 110
      - 115
      - 105
      - 111
      - 110
      - 115
      - 32
      - 99
      - 97
      - 110
      - 32
      - 116
      - 104
      - 101
      - 111
      - 114
      - 101
      - 116
      - 105
      - 99
      - 97
      - 108
      - 108
      - 121
      - 32
      - 98
      - 101
      - 32
      - 99
      - 104
      - 97
      - 110
      - 103
      - 101
      - 100
      - 44
      - 32
      - 98
      - 117
      - 116
      - 32
      - 121
      - 111
      - 117
      - 39
      - 108
      - 108
      - 32
      - 110
      - 101
      - 101
      - 100
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 32
      - 116
      - 111
      - 32
      - 109
      - 97
      - 110
      - 117
      - 97
      - 108
      - 108
      - 121
      - 32
      - 114
      - 101
      - 115
      - 105
      - 122
      - 101
      - 32
      - 116
      - 104
      - 101
      - 32
      - 118
      - 97
      - 114
      - 105
      - 111
      - 117
      - 115
      - 32
      - 112
      - 105
      - 99
      - 116
      - 117
      - 114
      - 101
      - 32
      - 98
      - 111
      - 120
      - 101
      - 115
      - 32
      - 116
      - 111
      - 32
      - 109
      - 97
      - 116
      - 99
      - 104
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - ConstKeyword:
      - 67
      - 111
      - 110
      - 115
      - 116
  - Whitespace:
      - 32
  - Identifier:
      - 87
      - 79
      - 82
      - 76
      - 68
      - 87
      - 73
      - 68
      - 84
      - 72
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Whitespace:
      - 32
  - EqualityOperator:
      - 61
  - Whitespace:
      - 32
  - Number:
      - 52
      - 57
      - 57
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - ConstKeyword:
      - 67
      - 111
      - 110
      - 115
      - 116
  - Whitespace:
      - 32
  - Identifier:
      - 87
      - 79
      - 82
      - 76
      - 68
      - 72
      - 69
      - 73
      - 71
      - 72
      - 84
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Whitespace:
      - 32
  - EqualityOperator:
      - 61
  - Whitespace:
      - 32
  - Number:
      - 52
      - 57
      - 57
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 65
      - 109
      - 111
      - 117
      - 110
      - 116
      - 32
      - 111
      - 102
      - 32
      - 102
      - 111
      - 111
      - 100
      - 32
      - 116
      - 111
      - 32
      - 103
      - 101
      - 110
      - 101
      - 114
      - 97
      - 116
      - 101
      - 32
      - 101
      - 118
      - 101
      - 114
      - 121
      - 32
      - 114
      - 111
      - 117
      - 110
      - 100
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 102
      - 111
      - 111
      - 100
      - 71
      - 101
      - 110
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 72
      - 111
      - 119
      - 32
      - 109
      - 117
      - 99
      - 104
      - 32
      - 101
      - 110
      - 101
      - 114
      - 103
      - 121
      - 32
      - 116
      - 104
      - 101
      - 32
      - 102
      - 111
      - 111
      - 100
      - 32
      - 103
      - 114
      - 97
      - 110
      - 116
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 102
      - 111
      - 111
      - 100
      - 87
      - 111
      - 114
      - 116
      - 104
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 68
      - 101
      - 102
      - 97
      - 117
      - 108
      - 116
      - 32
      - 115
      - 116
      - 97
      - 114
      - 116
      - 105
      - 110
      - 103
      - 32
      - 101
      - 110
      - 101
      - 114
      - 103
      - 121
      - 32
      - 102
      - 111
      - 114
      - 32
      - 101
      - 97
      - 99
      - 104
      - 32
      - 99
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 115
      - 116
      - 97
      - 114
      - 116
      - 69
      - 110
      - 101
      - 114
      - 103
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 82
      - 101
      - 103
      - 101
      - 110
      - 101
      - 114
      - 97
      - 116
      - 101
      - 32
      - 102
      - 111
      - 111
      - 100
      - 32
      - 101
      - 118
      - 101
      - 114
      - 121
      - 32
      - 40
      - 102
      - 111
      - 111
      - 100
      - 82
      - 101
      - 103
      - 101
      - 110
      - 41
      - 32
      - 110
      - 117
      - 109
      - 98
      - 101
      - 114
      - 32
      - 111
      - 102
      - 32
      - 116
      - 117
      - 114
      - 110
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 102
      - 111
      - 111
      - 100
      - 82
      - 101
      - 103
      - 101
      - 110
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 77
      - 117
      - 116
      - 97
      - 116
      - 105
      - 111
      - 110
      - 32
      - 118
      - 97
      - 108
      - 117
      - 101
      - 115
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 87
      - 104
      - 101
      - 116
      - 104
      - 101
      - 114
      - 32
      - 111
      - 114
      - 32
      - 110
      - 111
      - 116
      - 32
      - 116
      - 111
      - 32
      - 97
      - 108
      - 108
      - 111
      - 119
      - 32
      - 114
      - 101
      - 112
      - 114
      - 111
      - 100
      - 117
      - 99
      - 116
      - 105
      - 111
      - 110
      - 32
      - 40
      - 97
      - 46
      - 107
      - 46
      - 97
      - 46
      - 32
      - 116
      - 104
      - 101
      - 32
      - 34
      - 99
      - 101
      - 108
      - 105
      - 98
      - 97
      - 99
      - 121
      - 34
      - 32
      - 118
      - 97
      - 114
      - 105
      - 97
      - 98
      - 108
      - 101
      - 41
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 116
      - 111
      - 77
      - 117
      - 108
      - 116
      - 105
      - 112
      - 108
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - BooleanKeyword:
      - 66
      - 111
      - 111
      - 108
      - 101
      - 97
      - 110
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 72
      - 111
      - 119
      - 32
      - 109
      - 97
      - 110
      - 121
      - 32
      - 116
      - 117
      - 114
      - 110
      - 115
      - 32
      - 109
      - 117
      - 115
      - 116
      - 32
      - 112
      - 97
      - 115
      - 115
      - 32
      - 98
      - 101
      - 102
      - 111
      - 114
      - 101
      - 32
      - 99
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
      - 115
      - 32
      - 97
      - 114
      - 101
      - 32
      - 97
      - 108
      - 108
      - 111
      - 119
      - 101
      - 100
      - 32
      - 116
      - 111
      - 32
      - 114
      - 101
      - 112
      - 114
      - 111
      - 100
      - 117
      - 99
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 109
      - 117
      - 116
      - 97
      - 116
      - 101
      - 84
      - 117
      - 114
      - 110
      - 115
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 72
      - 111
      - 119
      - 32
      - 109
      - 97
      - 110
      - 121
      - 32
      - 112
      - 111
      - 116
      - 101
      - 110
      - 116
      - 105
      - 97
      - 108
      - 32
      - 109
      - 117
      - 116
      - 97
      - 116
      - 105
      - 111
      - 110
      - 115
      - 32
      - 116
      - 111
      - 32
      - 99
      - 97
      - 117
      - 115
      - 101
      - 32
      - 97
      - 116
      - 32
      - 114
      - 101
      - 112
      - 114
      - 111
      - 100
      - 117
      - 99
      - 116
      - 105
      - 111
      - 110
      - 32
      - 116
      - 105
      - 109
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 110
      - 117
      - 109
      - 79
      - 102
      - 77
      - 117
      - 116
      - 97
      - 116
      - 105
      - 111
      - 110
      - 115
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 65
      - 109
      - 111
      - 117
      - 110
      - 116
      - 32
      - 111
      - 102
      - 32
      - 99
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
      - 115
      - 32
      - 116
      - 111
      - 32
      - 115
      - 116
      - 97
      - 114
      - 116
      - 32
      - 119
      - 105
      - 116
      - 104
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 73
      - 110
      - 105
      - 116
      - 105
      - 97
      - 108
      - 67
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
      - 115
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 102
      - 111
      - 111
      - 100
      - 32
      - 97
      - 114
      - 114
      - 97
      - 121
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 70
      - 111
      - 111
      - 100
  - LeftParanthesis:
      - 40
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 87
      - 104
      - 101
      - 116
      - 104
      - 101
      - 114
      - 32
      - 111
      - 114
      - 32
      - 110
      - 111
      - 116
      - 32
      - 116
      - 111
      - 32
      - 100
      - 114
      - 97
      - 119
      - 32
      - 100
      - 101
      - 99
      - 101
      - 97
      - 115
      - 101
      - 100
      - 32
      - 99
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 100
      - 114
      - 97
      - 119
      - 68
      - 101
      - 97
      - 100
      - 67
      - 114
      - 101
      - 97
      - 116
      - 117
      - 114
      - 101
      - 115
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - BooleanKeyword:
      - 66
      - 111
      - 111
      - 108
      - 101
      - 97
      - 110
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 76
      - 111
      - 111
      - 112
      - 32
      - 118
      - 97
      - 114
      - 105
      - 97
      - 98
      - 108
      - 101
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 105
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 106
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 121
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
